### PR TITLE
feature(generator): support omitting rpcs from code generation

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6236)

### CODE REVIEW

**Comment by scotthart** [comment]
- Added support for command line argument to omit an rpc from the code emitted by the generator.
- Refactored ServiceCodeGenerator to be the sole source for the set of methods to have code emitted.
- Relocated tests for Has* methods that were moved into ServiceCodeGenerator.

File: generator/internal/codegen_utils.cc [hunk 1]
```diff
@@ -32,6 +33,81 @@ std::vector<std::pair<std::string, std::string>> const& SnakeCaseExceptions() {
...
+    std::vector<std::pair<std::string, std::string>>& command_line_args) {
+  auto copyright_year =
+      std::find_if(command_line_args.begin(), command_line_args.end(),
+                   [](std::pair<std::string, std::string> const& p) {
+                     return p.first == "copyright_year";
+                   });
+  if (copyright_year == command_line_args.end()) {
+    command_line_args.emplace_back("copyright_year", CurrentCopyrightYear());
+  } else if (copyright_year->second.empty()) {
+    copyright_year->second = CurrentCopyrightYear();
```
Comment (devjgm) [inline generator/internal/codegen_utils.cc:85]: Could `copyright_year` appear multiple times in these args?

File: generator/internal/codegen_utils.cc [hunk 2]
```diff
@@ -32,6 +33,81 @@ std::vector<std::pair<std::string, std::string>> const& SnakeCaseExceptions() {
...
+                           });
+  while (iter != command_line_args.end()) {
+    omitted_rpcs.insert(iter->second);
+    command_line_args.erase(iter);
+    iter = std::find_if(command_line_args.begin(), command_line_args.end(),
+                        [](std::pair<std::string, std::string> const& p) {
+                          return p.first == "omit_rpc";
+                        });
+  }
+  if (!omitted_rpcs.empty()) {
```
Comment (devjgm) [inline generator/internal/codegen_utils.cc:104]: IIUC, this code is basically replacing a potential sequence of `omit_rpc=foo` flags with a single `omitted_rpcs=foo1,foo2` flag. If that's correct, can we avoid this by just requiring that the caller provide the `ommitted_rpcs` flag in the first place?

**Review by devjgm — APPROVED** [review]

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6236)* status: 0 of 17 files reviewed, 2 unresolved discussions (waiting on @devjgm)

___
*[generator/internal/codegen_utils.cc, line 85 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6236#-MY86ga-DxV6o6Hnri94-r1-85:-MYC77SF5waMixjTm0q3:b-lme1f0) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0d394f43a740ae42b104ad8ad4d1630c9f4c6839/generator/internal/codegen_utils.cc#L85)):*
<details><summary><i>Previously, devjgm (Greg Miller) wrote…</i></summary><blockquote>

Could `copyright_year` appear multiple times in these args?
</blockquote></details>

Technically yes, but we would only user the first one encountered.
___
*[generator/internal/codegen_utils.cc, line 104 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6236#-MY86ga-DxV6o6Hnri94-r1-104:-MYC7G4RB0eSkAah4nGg:bhoms7) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0d394f43a740ae42b104ad8ad4d1630c9f4c6839/generator/internal/codegen_utils.cc#L104)):*
<details><summary><i>Previously, devjgm (Greg Miller) wrote…</i></summary><blockquote>

IIUC, this code is basically replacing a potential sequence of `omit_rpc=foo` flags with a single `omitted_rpcs=foo1,foo2` flag. If that's correct, can we avoid this by just requiring that the caller provide the `ommitted_rpcs` flag in the first place?
</blockquote></details>

I tried this first. The protoc command line argument code doesn't handle the presence of commas in the fashion that this would require.

**Comment by devjgm** [comment]
On Tue, Apr 13, 2021 at 5:28 PM Scott Hart ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> *Reviewable
> <https://reviewable.io/reviews/googleapis/google-cloud-cpp/6236>* status:
> 0 of 17 files reviewed, 2 unresolved discussions (waiting on @devjgm
> <https://github.com/devjgm>)
>
> ------------------------------
>
> *generator/internal/codegen_utils.cc, line 85 at r1
> <https://reviewable.io/reviews/googleapis/google-cloud-cpp/6236#-MY86ga-DxV6o6Hnri94-r1-85:-MYC77SF5waMixjTm0q3:b-lme1f0>
> (raw file
> <https://github.com/googleapis/google-cloud-cpp/blob/0d394f43a740ae42b104ad8ad4d1630c9f4c6839/generator/internal/codegen_utils.cc#L85>):*
> *Previously, devjgm (Greg Miller) wrote…*
>
> Could copyright_year appear multiple times in these args?
>
> Technically yes, but we would only user the first one encountered.
> ------------------------------
>
> *generator/internal/codegen_utils.cc, line 104 at r1
> <https://reviewable.io/reviews/googleapis/google-cloud-cpp/6236#-MY86ga-DxV6o6Hnri94-r1-104:-MYC7G4RB0eSkAah4nGg:bhoms7>
> (raw file
> <https://github.com/googleapis/google-cloud-cpp/blob/0d394f43a740ae42b104ad8ad4d1630c9f4c6839/generator/internal/codegen_utils.cc#L104>):*
> *Previously, devjgm (Greg Miller) wrote…*
>
> IIUC, this code is basically replacing a potential sequence of
> omit_rpc=foo flags with a single omitted_rpcs=foo1,foo2 flag. If that's
> correct, can we avoid this by just requiring that the caller provide the
> ommitted_rpcs flag in the first place?
>
> I tried this first. The protoc command line argument code doesn't handle
> the presence of commas in the fashion that this would require.
>
Can you use something other than commas then? Semicolon? Hyphen?

**Pull request merged**